feat: append keploy version on each test case and mock stored in yaml files

besasch88 opened the pull request:
…

## Related Issue
  - Append to each test and mock the Keploy version to facilitate the debugging

Closes: https://github.com/keploy/keploy/issues/1848 

#### Describe the changes you've made
A clear and concise description of what you have done to successfully close your assigned issue. Any new files? or anything you feel to let us know!

## Type of change

<!--
Example how to mark a checkbox:-
- [x] My code follows the code style of this project.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Please let us know if any test cases are added

Please describe the tests(if any). Provide instructions how its affecting the coverage.

#### Describe if there is any unusual behaviour of your code(Write `NA` if there isn't)
A clear and concise description of it.

## Checklist:
<!--
Example how to mark a checkbox:-
- [x] My code follows the code style of this project.
-->
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.

## Screenshots (if any)

Here some examples of generated YAML Files:

Test:
![Screenshot 2024-06-26 at 21 44 01](https://github.com/keploy/keploy/assets/29631972/7557d8aa-88ef-4fcf-b9da-a5355464b9dc)

Mock:
![Screenshot 2024-06-26 at 21 44 17](https://github.com/keploy/keploy/assets/29631972/b53c1d0f-2a30-4ae8-9a31-7b9037d77142)

Multiple mocks in same file:
![Screenshot 2024-06-26 at 21 44 21](https://github.com/keploy/keploy/assets/29631972/0863ae38-bef5-4177-9ad9-781d703eeae6)

besasch88 commented:
**I have read the CLA Document and I hereby sign the CLA**

gouravkrosx commented:
@besasch88 In the case of mocks also it should be only once just like tests, at the top and not for the individual mock.

besasch88 commented:
> @besasch88 In the case of mocks also it should be only once just like tests, at the top and not for the individual mock.

Thanks @gouravkrosx! Looking at the code, where I added the logic, we are within the scope of the single mock, it is the yaml package that writes each mock into the existing file by concatenating it. To meet your proposal, it means we have to move this logic inside the `yaml` package, where we have information about the mocks already existing in the yaml files (please see the example code below). I would honestly avoid moving the logic there, because it means adding custom logic to a generic package that could be leveraged in the future for other areas.


Here the example code we should add in `pkg/platform/yaml/yaml.go:66` like this:
```go
func WriteFile(ctx context.Context, logger *zap.Logger, path, fileName string, docData []byte, isAppend bool) error {
     isFileEmpty, err := CreateYamlFile(ctx, logger, path, fileName)
     if err != nil {
        return err
     }
+   if isFileEmpty {
+      docData = append([]byte(utils.GenerateKeployVersionComment()), docData...)
+   }
     flag := os.O_WRONLY | os.O_TRUNC
     ...
``` 

Like to get your opinion around it, I believe mainly depends on the best practices on how define the boundaries of each package

gouravkrosx commented:
Can't we stat the file and see if it exists or not, if not then only prepend the data in case of mocks?
And can you show me other files as well like the report, and config where this comment can be seen?

besasch88 commented:
> Can't we stat the file and see if it exists or not, if not then only prepend the data in case of mocks? And can you show me other files as well like the report, and config where this comment can be seen?

@gouravkrosx, I updated the PR. Basically for those who call `yaml.WriteFile` with the append parameter to true, I check if the file exists, if not I add the comment. This is valid for mocks, for others the comment is always added because the file is created/replaced.  

I added the same comment for the `keploy config --generate` command
| Config: ![config](https://github.com/keploy/keploy/assets/29631972/24c218f9-8188-4fbb-8963-1e0152ec9f26) | Test: ![test](https://github.com/keploy/keploy/assets/29631972/31b19075-c96a-4b5a-8c6e-da86a254e313) |
|--------|--------|
|Mock: ![mock](https://github.com/keploy/keploy/assets/29631972/764de914-cc1c-46c2-bf9b-16077e40391c)| Appended Mock: ![mock_append](https://github.com/keploy/keploy/assets/29631972/9404796d-bc52-4dd8-9478-725e7df824fa) |

gouravkrosx commented on utils/utils.go:
```diff
@@ -213,6 +213,10 @@ func CheckFileExists(path string) bool {
 
 var Version string
 
+func GenerateKeployVersionComment() string {
```
Just rename it to GetVersionComment()

besasch88 replied:
Done! ;)

gouravkrosx commented:
@besasch88 I will try to review this PR by tomorrow EOD.

PranshuSrivastava commented:
Hey @besasch88, please fix the merge conflicts so that this PR can be reviewed.

PranshuSrivastava commented:
Closing as this is stale
